### PR TITLE
feat(scores): add non-numeric scores and score-configs via API, drop non-null cond on `value` in `scores` table

### PR DESCRIPTION
**Pull request by marliessophie**

### CODE REVIEW

File: packages/shared/src/features/annotation/types.ts [hunk 1]
```diff
@@ -13,3 +13,92 @@ export type ConfigCategory = z.infer<typeof configCategory>;
...
+  authorUserId: z.string().nullish(),
+  comment: z.string().nullish(),
+  traceId: z.string(),
+  observationId: z.string().nullish(),
+  configId: z.string().nullish(),
+  createdAt: z.date(),
+  updatedAt: z.date(),
+});
+
+const NumericData = z.object({
```
Comment (marcklingen) [inline packages/shared/src/features/annotation/types.ts:34]: you could use ScoreBase.extend() here which would reduce the need for merge below and make NumericData usable independently if necessary later

File: packages/shared/src/features/annotation/types.ts [hunk 2]
```diff
@@ -13,3 +13,92 @@ export type ConfigCategory = z.infer<typeof configCategory>;
...
+  comment: z.string().nullish(),
+  traceId: z.string(),
+  observationId: z.string().nullish(),
+  trace: z.object({
+    userId: z.string(),
+  }),
+});
+
+export const GetAllScores = z.discriminatedUnion("dataType", [
+  GetAllScoresBase.merge(NumericData),
```
Comment (marcklingen) [inline packages/shared/src/features/annotation/types.ts:72]: how dies getall work with value but without an operator?

File: web/src/features/manual-scoring/components/AnnotateDrawer.tsx [hunk 3]
```diff
@@ -301,47 +303,67 @@ export function AnnotateDrawer({
...
+              capture("score:create", {
+                type: type,
+                source: source,
+                dataType: score.dataType,
+              });
+            }
           }
         }
+      } catch (error) {
+        trpcErrorToast(error);
```
Comment (marcklingen) [inline web/src/features/manual-scoring/components/AnnotateDrawer.tsx:366]: trpcErrorToast(error) does not need to be passed manually or onError above, it is used automatically as onError for all trpc routes unless it is overridden by onError of a mutation or query

File: web/src/features/prompts/server/routers/promptRouter.ts [hunk 4]
```diff
@@ -635,6 +635,7 @@ export const promptRouter = createTRPCRouter({
           WHERE
               o.type = 'GENERATION'
               AND s.data_type != 'CATEGORICAL'
+              AND s.value IS NOT NULL
```
Comment (marcklingen) [inline web/src/features/prompts/server/routers/promptRouter.ts:638]: do you have an issue with quick follow ons regarding this (need for displaying categorical scores (value counts) on tables

File: packages/shared/src/features/annotation/types.ts (hunk 2 above)
Comment (marliessophie) [reply packages/shared/src/features/annotation/types.ts]: I don't understand your question, could you rephrase please?

File: web/src/features/prompts/server/routers/promptRouter.ts (hunk 4 above)
Comment (marliessophie) [reply web/src/features/prompts/server/routers/promptRouter.ts:638]: Yes: https://linear.app/langfuse/issue/LFE-1270/add-categorical-score-counts-averages-in-charts-prompt-metrics

File: packages/shared/src/features/annotation/types.ts [hunk 5]
```diff
@@ -1,15 +1,97 @@
...
+  updatedAt: z.coerce.date(),
+});
+
+export const ValidatedScoreSchema = z.discriminatedUnion("dataType", [
+  ScoreBase.merge(NumericData),
+  ScoreBase.merge(CategoricalData),
+  ScoreBase.merge(BooleanData),
+]);
+
+export type ValidatedScore = z.infer<typeof ValidatedScoreSchema>;
```
Comment (marliessophie) [inline packages/shared/src/features/annotation/types.ts:72]: Note: this currently is defined once for public api and once here for langfuse codebase use. Is there a better way to do this? Ideally I'd like to share this type across both use cases, to ensure we always expect scores objects to have this data type

File: packages/shared/src/features/annotation/types.ts [hunk 6]
```diff
@@ -1,15 +1,86 @@
...
+  configId: z.string().nullish(),
+  createdAt: z.coerce.date(),
+  updatedAt: z.coerce.date(),
+});
+
+export const ValidatedScoreSchema = z.discriminatedUnion("dataType", [
+  ScoreBase.merge(NumericData),
+  ScoreBase.merge(CategoricalData),
+  ScoreBase.merge(BooleanData),
+]);
```
Comment (marliessophie) [inline packages/shared/src/features/annotation/types.ts:59]: Note: this currently is defined once for public api and once here for langfuse codebase use. Is there a better way to do this? Ideally I'd like to share this type across both use cases, to ensure we always expect scores objects to have this data type

File: web/src/features/public-api/types/score-configs.ts [hunk 7]
```diff
@@ -0,0 +1,190 @@
+import {
+  jsonSchema,
+  paginationMetaResponseZod,
+  paginationZod,
+} from "@langfuse/shared";
+import { z } from "zod";
+
+const isPresent = <T>(value: T): value is NonNullable<T> =>
+  value !== null && value !== undefined;
```
Comment (marliessophie) [inline web/src/features/public-api/types/score-configs.ts:9]: Do we have some place in the codebase for generic helpers? I'd like to move this function there, I think it's helpful in general. For example if a numeric 0 is a valid input, a falsy check would invalidate it 🤷

File: packages/shared/src/features/annotation/types.ts (hunk 1 above)
Comment (marliessophie) [reply packages/shared/src/features/annotation/types.ts:22]: I think the current version is more readable, but feel free to disagree :)

File: fern/apis/client/definition/score.yml [hunk 8]
```diff
@@ -17,15 +17,100 @@ types:
...
+      source: ScoreSource
+      observationId: optional<string>
+      timestamp: datetime
+      comment: optional<string>
+      dataType: ScoreDataTypeCategorical
+      configId:
+        type: optional<string>
+        docs: Reference a score config on a score. When set, config and score name must be equal and stringValue must map to a config category
+  Score:
+    discriminated: false
```
Comment (marcklingen) [inline fern/apis/client/definition/score.yml:86]: could we use discriminated here?

Comment (marcklingen) [inline fern/apis/client/definition/score.yml:86]: could we use discriminated here?

Comment (marcklingen) [reply fern/apis/client/definition/score.yml:86]: + literal for the type instead of enums with a signle value to keep the openapi spec simpler

File: fern/apis/server/definition/commons.yml [hunk 9]
```diff
@@ -120,25 +120,89 @@ types:
...
       name: string
-      value: double
+      value:
+        type: double
+        docs: The numeric value of the score
+      source: ScoreSource
+      observationId: optional<string>
+      timestamp: datetime
+      comment: optional<string>
+      dataType: ScoreDataTypeNumeric
```
Comment (marcklingen) [inline fern/apis/server/definition/commons.yml:152]: see notes above on using discriminated type with a literal instead of enums with a single value

File: fern/apis/server/definition/commons.yml [hunk 10]
```diff
@@ -120,25 +120,89 @@ types:
...
       source: ScoreSource
       observationId: optional<string>
       timestamp: datetime
       comment: optional<string>
+      dataType: ScoreDataTypeCategorical
+      configId:
+        type: optional<string>
+        docs: Reference a score config on a score. When set, config and score name must be equal and stringValue must map to a config category
+  Score:
+    discriminated: false
```
Comment (marcklingen) [inline fern/apis/server/definition/commons.yml:195]: see above, i'd go for discriminated type here

File: fern/apis/server/definition/ingestion.yml [hunk 11]
```diff
@@ -185,9 +185,32 @@ types:
...
+        docs: The value of the score. Must be passed as string for categorical scores, and numeric for boolean and numeric scores. Boolean score values must equal either 1 or 0 (true or false)
       observationId: optional<string>
       comment: optional<string>
+      dataType:
+        type: optional<commons.ScoreDataType>
+        docs: When set, must match the score value's type. If not set, will be inferred from the score value or config
+      configId:
+        type: optional<string>
+        docs: Reference a score config on a score. When set, the score name must equal the config name and scores must comply with the config's range and data type. For categorical scores, the value must map to a config category. Numeric scores might be constrained by the score config's max and min values
+    docs: |
```
Comment (marcklingen) [inline fern/apis/server/definition/ingestion.yml:199]: I like the examples, should add on post /scores as well

Comment (marcklingen) [reply fern/apis/server/definition/ingestion.yml:199]: post /scores is more used via api than sdks, /ingestion only relevant for users building own tracing clients

File: packages/shared/src/features/annotation/types.ts (hunk 6 above)
Comment (marcklingen) [reply packages/shared/src/features/annotation/types.ts:59]: mh, if this is a public api interface, I'd move it to feature/public-api/types and then import elsewhere. Thereby it is clear that this is an external interface that must not be changed / covered by semver.

File: packages/shared/src/features/ingestion/types.ts [hunk 12]
```diff
@@ -145,15 +145,44 @@ export const UpdateGenerationBody = UpdateSpanBody.extend({
...
   traceId: z.string(),
   observationId: z.string().nullish(),
   comment: z.string().nullish(),
+  configId: z.string().nullish(),
 });
 
+export const ScoreBody = z.discriminatedUnion("dataType", [
+  BaseScoreBody.merge(
+    z.object({
+      value: z.number(),
```
Comment (marcklingen) [inline packages/shared/src/features/ingestion/types.ts:160]: I'd add custom zod errors here as this might be the biggest issue when folks use the api, e.g. "value must be numeric for numeric scores"

File: packages/shared/src/features/ingestion/types.ts [hunk 13]
```diff
@@ -145,15 +145,44 @@ export const UpdateGenerationBody = UpdateSpanBody.extend({
...
+  BaseScoreBody.merge(
+    z.object({
+      value: z.string(),
+      dataType: z.literal("CATEGORICAL"),
+    })
+  ),
+  BaseScoreBody.merge(
+    z.object({
+      value: z.number().refine((val) => val === 0 || val === 1, {
+        message: "Value must be either 0 or 1",
```
Comment (marcklingen) [inline packages/shared/src/features/ingestion/types.ts:173]: see above, reference BOOLEAN datatype

File: web/src/__tests__/score-configs.servertest.ts [hunk 14]
```diff
@@ -131,13 +144,242 @@ describe("/api/public/score-configs API Endpoint", () => {
...
+      "POST",
+      "/api/public/score-configs",
+      {
+        name: "numeric-config-name",
+        dataType: "NUMERIC",
+        maxValue: 0,
+      },
+    );
+
+    const dbScoreConfig = await prisma.scoreConfig.findMany({
```
Comment (marcklingen) [inline web/src/__tests__/score-configs.servertest.ts:170]: i'd probably use the get api for all of these tests instead of prisma to make sure the api works with all of the different types

File: web/src/__tests__/scores.servertest.ts [hunk 15]
```diff
@@ -136,43 +129,542 @@ describe("/api/public/scores API Endpoint", () => {
...
-      release: "1.0.0",
-      version: "2.0.0",
+    const scoreId = uuidv4();
+    const scoreData = {
+      id: scoreId,
+      name: "accuracy",
+      value: "Good",
+      traceId,
+    };
+    const createScore = await makeAPICall(
```
Comment (marcklingen) [inline web/src/__tests__/scores.servertest.ts:208]: typed api call here?

File: web/src/__tests__/scores.servertest.ts [hunk 16]
```diff
@@ -136,43 +129,542 @@ describe("/api/public/scores API Endpoint", () => {
...
+
+    const scoreId = uuidv4();
+    const scoreData = {
+      id: scoreId,
+      name: "score-name",
+      value: 1,
+      dataType: "BOOLEAN",
+      traceId,
+    };
+    const createScore = await makeAPICall(
```
Comment (marcklingen) [inline web/src/__tests__/scores.servertest.ts:247]: typed api call here?

File: web/src/__tests__/scores.servertest.ts [hunk 17]
```diff
@@ -136,43 +129,542 @@ describe("/api/public/scores API Endpoint", () => {
...
+      "POST",
+      "/api/public/scores",
+      scoreData,
+    );
+
+    expect(createScore.status).toBe(200);
+    const dbScore = await prisma.score.findUnique({
```
Comment (marcklingen) [inline web/src/__tests__/scores.servertest.ts:254]: fetch via api to test the get method for all of these against the schema?

File: web/src/__tests__/scores.servertest.ts [hunk 18]
```diff
@@ -354,13 +835,28 @@ describe("/api/public/scores API Endpoint", () => {
       await makeAPICall("POST", "/api/public/generations", {
         id: generationId,
       });
+
+      await makeAPICall("POST", "/api/public/score-configs", {
```
Comment (marcklingen) [inline web/src/__tests__/scores.servertest.ts:839]: used typed call here

File: web/src/features/datasets/server/dataset-router.ts [hunk 19]
```diff
@@ -620,6 +626,26 @@ export const datasetRouter = createTRPCRouter({
         `,
       );
 
+      const validatedTraceScores = traceScores.reduce((acc, ts) => {
```
Comment (marcklingen) [inline web/src/features/datasets/server/dataset-router.ts:629]: does it make sense to keep a function to transorm prisma score into zod-score at a central place? similar to how it is done for the models api

File: web/src/features/manual-scoring/components/CreateScoreConfigButton.tsx [hunk 20]
```diff
@@ -35,30 +34,62 @@ import {
...
   minValue: z.coerce.number().optional(),
   maxValue: z.coerce.number().optional(),
-  categories: z.array(category).optional(),
+  categories: z.array(Category).optional(),
   description: z.string().optional(),
 });
 
+type CreateConfig = z.infer<typeof createConfigSchema>;
+
+const validateScoreConfig = (values: CreateConfig): string | null => {
```
Comment (marcklingen) [inline web/src/features/manual-scoring/components/CreateScoreConfigButton.tsx:54]: see above, does it make sense to keep this check in a central place and use for ui, trpc route and api?

File: web/src/features/public-api/types/score-configs.ts (hunk 7 above)
Comment (marcklingen) [reply web/src/features/public-api/types/score-configs.ts:9]: yes, utils

File: web/src/features/public-api/types/score-configs.ts [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+      ) {
+        ctx.addIssue({
+          code: z.ZodIssueCode.custom,
+          message: "Maximum value must be greater than Minimum value",
+        });
+      }
+    }
+  });
+
+// I need this type on the frontend as well, how to best handle
```
Comment (marcklingen) [inline web/src/features/public-api/types/score-configs.ts:126]: lets chat about this

File: web/src/features/public-api/types/scores.ts [hunk 22]
```diff
@@ -0,0 +1,244 @@
+import { isPresent } from "@/src/features/manual-scoring/lib/helpers";
```
Comment (marcklingen) [inline web/src/features/public-api/types/scores.ts:1]: make this the source of truth?

File: web/src/pages/api/public/score-configs.ts [hunk 23]
```diff
@@ -1,109 +1,119 @@
...
-const ScoreConfigsGetSchema = z.object({
-  ...paginationZod,
-});
-
-export default async function handler(
-  req: NextApiRequest,
-  res: NextApiResponse,
-) {
-  await runMiddleware(req, res, cors);
+const inflateConfigBody = (body: z.infer<typeof PostScoreConfigBody>) => {
```
Comment (marcklingen) [inline web/src/pages/api/public/score-configs.ts:17]: move this to types?

File: web/src/pages/api/public/score-configs.ts [hunk 24]
```diff
@@ -1,109 +1,119 @@
...
+      const config = await prisma.scoreConfig.create({
+        data: {
+          ...inflatedConfigInput,
+          categories: inflatedConfigInput.categories ?? undefined,
+          id: v4(),
+          projectId: auth.scope.projectId,
+        },
+      });
 
+      return config as z.infer<typeof ScoreConfig>;
```
Comment (marcklingen) [inline web/src/pages/api/public/score-configs.ts:61]: the goal of this api wrapper is to throw type warnings in case the underlying db types drift from the api type. by defining the return type as this type, this type warning does not work anymore

File: web/src/pages/api/public/score-configs.ts [hunk 25]
```diff
@@ -1,109 +1,119 @@
...
-          message: "Internal Server Error",
-          error: "Invalid config format encountered",
-        });
-      }
+      const configs = rawConfigs.reduce(
+        (acc, config) => {
+          if (ScoreConfig.safeParse(config).success) {
+            acc.push(config as z.infer<typeof ScoreConfig>);
+          } else {
+            Sentry.captureException(
```
Comment (marcklingen) [inline web/src/pages/api/public/score-configs.ts:86]: i like

File: web/src/pages/api/public/score-configs.ts [hunk 26]
```diff
@@ -1,109 +1,119 @@
...
-    }
-
-    if (req.method === "GET") {
-      if (authCheck.scope.accessLevel !== "all") {
-        return res.status(401).json({
-          message: "Access denied - need to use basic auth with secret key",
-        });
+      if (existingConfig) {
+        throw new InvalidRequestError(
+          "Score config with this name and data type already exists for this project",
```
Comment (marcklingen) [inline web/src/pages/api/public/score-configs.ts:46]: should they be unique on name or on name/type?

File: web/src/pages/api/public/score-configs.ts (hunk 23 above)
Comment (marcklingen) [reply web/src/pages/api/public/score-configs.ts:17]: i like this to make booleans simpler, great idea

File: web/src/pages/api/public/scores.ts [hunk 27]
```diff
@@ -1,167 +1,129 @@
...
+  PostScoresBody,
+  PostScoresResponse,
+  type ValidatedGetScoresData,
+} from "@/src/features/public-api/types/scores";
 
+export default withMiddlewares({
+  POST: createAuthedAPIRoute({
+    name: "Create Score",
+    bodySchema: PostScoresBody,
+    responseSchema: PostScoresResponse,
```
Comment (marcklingen) [inline web/src/pages/api/public/scores.ts:25]: as this routes via the batch endpoint, we do not get a scores object back which is not nice. shall we add additional logic here?

File: web/src/pages/api/public/scores.ts [hunk 28]
```diff
@@ -1,167 +1,129 @@
...
           : Prisma.empty;
-      const scoreIdCondition = obj.scoreIds
-        ? Prisma.sql`AND s."id" = ANY(${obj.scoreIds})`
+      const scoreIdCondition = scoreIds
+        ? Prisma.sql`AND s."id" = ANY(${scoreIds})`
         : Prisma.empty;
 
       const scores = await prisma.$queryRaw<
-        Array<Score & { trace: { userId: string } }>
+        Array<ValidatedGetScoresData>
```
Comment (marcklingen) [inline web/src/pages/api/public/scores.ts:89]: we could type as unknown here as you afterwards validate

File: web/src/server/api/services/EventProcessor.ts [hunk 29]
```diff
@@ -606,6 +616,145 @@ export class ScoreProcessor implements EventProcessor {
     this.event = event;
   }
 
+  static inferDataType(value: string | number): ScoreDataType {
```
Comment (marcklingen) [inline web/src/server/api/services/EventProcessor.ts:619]: centralize these helpers?

File: fern/apis/client/definition/score.yml (hunk 8 above)
Comment (marcklingen) [reply fern/apis/client/definition/score.yml:86]: see IngestionEvent for an example

File: fern/apis/server/definition/ingestion.yml (hunk 11 above)
Comment (marcklingen) [reply fern/apis/server/definition/ingestion.yml:199]: https://buildwithfern.com/learn/api-definition/fern/examples#objects

File: web/src/pages/api/public/score-configs.ts (hunk 26 above)
Comment (marcklingen) [reply web/src/pages/api/public/score-configs.ts:46]: mutual decision: drop here. if we were to do this, then add in db

File: web/src/pages/api/public/scores.ts (hunk 27 above)
Comment (marcklingen) [reply web/src/pages/api/public/scores.ts:25]: ignore for now, could add later

Comment (marcklingen) [reply web/src/pages/api/public/scores.ts:25]: but cover in tests that it is void
